pkg/agent/billing.go

```diff
@@ -0,0 +1,246 @@
+package agent
+
+import (
+	"context"
+	"math"
+	"net/http"
+	"time"
+
+	"github.com/google/uuid"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+
+	vmapi "github.com/neondatabase/neonvm/apis/neonvm/v1"
+
+	"github.com/neondatabase/autoscaling/pkg/billing"
+	"github.com/neondatabase/autoscaling/pkg/util"
+)
+
+type BillingConfig struct {
+	URL                 string `json:"url"`
+	CPUMetricName       string `json:"cpuMetricName"`
+	CollectEverySeconds uint   `json:"collectEverySeconds"`
+	PushEverySeconds    uint   `json:"pushEverySeconds"`
+	PushTimeoutSeconds  uint   `json:"pushTimeoutSeconds"`
+}
+
+type billingMetricsState struct {
+	historical      map[billingMetricsKey]vmMetricsHistory
+	present         map[billingMetricsKey]vmMetricsInstant
+	lastCollectTime *time.Time
+	pushWindowStart time.Time
+}
+
+type billingMetricsKey struct {
+	uid        types.UID
+	endpointID string
+}
+
+type vmMetricsHistory struct {
+	lastSlice *metricsTimeSlice
+	total     vmMetricsSeconds
+}
+
+type metricsTimeSlice struct {
+	metrics   vmMetricsInstant
+	startTime time.Time
+	endTime   time.Time
+}
+
+type vmMetricsInstant struct {
+	// cpu stores the cpu allocation at a particular instant.
+	cpu uint16
+}
+
+// vmMetricsSeconds is like vmMetrics, but the values cover the allocation over time
+type vmMetricsSeconds struct {
+	// cpu stores the CPU seconds allocated to the VM, roughly equivalent to the integral of CPU
+	// usage over time.
+	cpu uint32
+}
+
+const (
+	EndpointLabel string = "neon/endpoint-id"
+)
+
+func RunBillingMetricsColllector(
+	backgroundCtx context.Context,
+	conf *BillingConfig,
+	targetNode string,
+	store *util.WatchStore[vmapi.VirtualMachine],
+) {
+	client := billing.NewClient(conf.URL, http.DefaultClient)
+
+	collectTicker := time.NewTicker(time.Second * time.Duration(conf.CollectEverySeconds))
+	defer collectTicker.Stop()
+	// Offset by half a second, so it's a bit more deterministic.
+	time.Sleep(500 * time.Millisecond)
+	pushTicker := time.NewTicker(time.Second * time.Duration(conf.PushEverySeconds))
+	defer pushTicker.Stop()
+
+	state := billingMetricsState{
+		historical:      make(map[billingMetricsKey]vmMetricsHistory),
+		present:         make(map[billingMetricsKey]vmMetricsInstant),
+		lastCollectTime: nil,
+		pushWindowStart: time.Now(),
+	}
+
+	state.collect(conf, targetNode, store)
+	batch := client.NewBatch()
+
+	for {
+		select {
+		case <-collectTicker.C:
+			klog.Infof("Collecting billing state")
+			state.collect(conf, targetNode, store)
+		case <-pushTicker.C:
+			klog.Infof("Creating billing batch")
+			state.drainAppendToBatch(conf, batch)
+			klog.Infof("Pushing billing events (count = %d)", batch.Count())
+			if err := pushBillingEvents(conf, batch); err != nil {
+				klog.Errorf("Error pushing billing events: %s", err)
+				continue
```

Comment:
if we get here, we have `n` things in the batch, the next time we get to line 98, we begin adding more things to the batch, correct, which means every time hit this condition, the batch will get bigger, thereby decreasing its chance of success. (particularly given the 500ms window and the chance that these could be running in different regions.)

Reply:
Yeah, it's not great. Realistically, size of JSON shouldn't cause us latency issues (as discussed, it'll still be quite small). Timeouts are worth considering, but also the configured duration is in seconds, so... :shrug:

When we add metrics, it'd be good to have something tracking the amount of time these requests take.